playbook/app/pb_kits/playbook/pb_passphrase/useHaveIBeenPwned.js

```diff
@@ -0,0 +1,51 @@
+
+import  { useEffect, useState } from 'react'
+
+const checkHaveIBeenPwned = async function (passphrase) {
+  const buffer = new TextEncoder('utf-8').encode(passphrase)
+  const digest = await crypto.subtle.digest('SHA-1', buffer)
+  const hashArray = Array.from(new Uint8Array(digest))
+  const hashHex = hashArray.map((b) => b.toString(16).padStart(2, '0')).join('')
+
+  const firstFive = hashHex.slice(0, 5)
+  const endOfHash = hashHex.slice(5)
+
+  const resp = await fetch(`https://api.pwnedpasswords.com/range/${firstFive}`)
+  const text = await resp.text()
+
+  const match = text.split('\n').some((line) => {
+    //Each line is <sha-1-hash-suffix>:<count of incidents>
+    return line.split(':')[0] === endOfHash.toUpperCase()
+  })
+  return match
+}
+
+/**
+ * If the input hasn't changed in <delay> ms,
+ * hit the haveibeenpwned api and check if the given passphrase is compromised
+ */
+export default function useHaveIBeenPwned(passphrase, minLength, delay = 400) {
+  const [isPwned, setIsPwned] = useState(false)
+
+  useEffect(
+    () => {
+      // only check the API for passphrases above the minimum size
+      if (passphrase.length < minLength) {
+        setIsPwned(false)
+        return
+      }
+
+      const handler = setTimeout(() => {
+        checkHaveIBeenPwned(passphrase)
+          .then((pwned) => setIsPwned(pwned))
```

Comment:
Since this hits an external web API, I think it makes sense to have some failover here incase the service is down or returning something other than 2xx code with a `.catch` and `.finally` which would set `pwned` to `false`. We will not know for sure whether to set the value to `true` in this case, so best to default this way.

Reply:
Great call. I was thinking "If the api is down, it can never be set to true" but that doesn't hold if it falls over after any request is successful. I'll set a default!